### ISSUE TITLE
generator exec logic might end up logging to journal, and really should not

### ISSUE THREAD

**Issue by biubiuzy** [systemd/systemd#35723]
### systemd version the issue has been seen with

255

### Used distribution

NA

### Linux kernel version used

6.10

### CPU architectures issue was seen on

arm

### Component

systemd

### Expected behaviour you didn't see

The systemd process is started successfully.

### Unexpected behaviour you saw

Failed to start the systemd process.
![image](https://github.com/user-attachments/assets/45f78fce-b094-407a-87f1-60ad2c38df99)

### Steps to reproduce the problem

The sd-gens process parses /etc/fstab and mounts the file system. When the log target is journal and the process prints a large number of logs, the sendmsg timeout (10s) is triggered. As a result, the systemd process fails to start after 90s.

I tried to modify the parse_fstab implementation locally to print more logs. However, the problem can be reproduced without other processing.

**Comment by JetXujing** [comment]
Maybe we can consider that logs of the switch root process cannot be printed to journals? For example, we can print to kmsg.

**Comment by biubiuzy** [comment]
The SELinux label is changed, and the permission to access the cgroup file is insufficient. As a result, fstab-generator opens journal and writes logs.

**Comment by bluca** [comment]
Have you reported this issue to the providers of your policy? This sounds like something that needs to be fixed in the policy

**Comment by JetXujing** [comment]
I think selinux is only a trigger scenario. The essence of the problem is that it is insecure to write logs to journals during switch-root (fstab-generator belongs to this period), because journald will be stopped before switch-root. I do not understand why logs are written to journald. Maybe we can set target to kmsg or console.

**Comment by bluca** [comment]
That's really not the case, the journal is widely used in the initrd and designed to do so. Have you tried to boot with selinux=0 on the kernel command line?

**Comment by JetXujing** [comment]
The initrd can print journal logs because systemd-journald runs successfully during the initrd process. There's no problem.

I mean, during the switch-root(initrd-switch-root.service), and I looked at the logs and found that the fstab-generate process took time during that process, and before fstab-generate, systemd-journald is killed by process 1 sending SIGTERM. During this period, logs written to journals are stored in the socket buffer. When the buffer is full, timeout is triggered. If the number of timeout times is too large, systemd will time out.

If selinux is set to 0, this problem can be avoided. This is because the following api will be invoked, and log_open_journal will not be invoked.
https://github.com/systemd/systemd/blob/2cc58b6c8ad1826f22b120c4148d123018a7b92c/src/shared/generator.c#L1001

**Comment by JetXujing** [comment]
What I'm thinking is why this uses LOG_TARGET_JOURNAL_OR_KMSG instead of LOG_TARGET_KMSG or other targets that do not contain journals.
https://github.com/systemd/systemd/blob/2cc58b6c8ad1826f22b120c4148d123018a7b92c/src/shared/generator.c#L1004

**Comment by bluca** [comment]
> If selinux is set to 0, this problem can be avoided. This is because the following api will be invoked, and log_open_journal will not be invoked.

Not really: it's not that the "problem is avoided", it's that the selinux policy is introducing a bug. Fix the policy, and everything will work as intended.

**Comment by poettering** [comment]
hmm, this might be a bug in pid1 code indeed. i.e. pid 1 only writes to the journal if it can verify the journal is up and running (it does so via an explicit check of the unit systemd-journald.service is active). This works great, but when we do the "exec-everything-on-some-dir" logic, then we'll fork an intermediary child and that child is not going to be pid1 so some special casing won't apply and we might end up logging to the journald even though conceptually we are part of pid 1 and really shouldn't do that. i.e. we should probably either stop reopening logging in that intermediary child or we should explicitly reopen with IPC-based logging off.

**Comment by bluca** [comment]
The error happens because the policy denies access, and it shouldn't

**Comment by JetXujing** [comment]
It is undeniable that this problem is caused by SELinux changes. SELinux should probably fix it, but it doesn't fundamentally solve the problem.
I always believe that SELinux is only one of the trigger scenarios. If many programs print logs to journals during switch-root (for example, debug is enabled), this problem still occurs. In this case, selinux is irrelevant.

**Comment by poettering** [comment]
So I looked at this again. So the special casing of PID 1 in the logging infra shouldn't be an issue after all. And yes, this all seems to be caused by selinux misconfig: we cannot determine whether we run in system context anymore.

Hence, this really needs to be fixed in selinux not in systemd.

That said, I prepped a "fix" for this nonetheless: instead of checking the cgroup it probably makes sense to check $SYSTEMD_SCOPE being set to "system" instead. That's how we communicate to generators in which context they are called, and they really should use that.

**Comment by bluca** [comment]
Regardless of this workaround, the fact that the selinux policy blocks generators from reading sysfs is incredibly wrong, and will likely break custom, local generators. Generators are supposed to be able to read whichever state they need to read to make decisions. Selinux stopping generators from reading sysfs is a bug in selinux that needs to be fixed.
